src/test/scala-2/chiselTests/simulator/PeekPokeApiSpec.scala

```diff
@@ -0,0 +1,282 @@
+          dut.io.in.valid.poke(false)
+
+          val peekedOp = dut.io.op.peek()
+          assert(peekedOp.litValue == op.litValue)
+          assert(peekedOp.toString.contains(TestOp.getClass.getSimpleName.stripSuffix("$")))
+
+          val expected = op match {
+            case TestOp.Add => a + b
+            case TestOp.Sub => a - b
+            case TestOp.Mul => a * b
+            case _          => throw new Exception("Invalid operation")
+          }
+          val expectedCmp = a.compare(b) match {
+            case -1 => dut.CmpResult.LT
+            case 0  => dut.CmpResult.EQ
+            case 1  => dut.CmpResult.GT
+          }
+
+          dut.io.out.valid.expect(true.B)
+          dut.io.out.valid.expect(true)
+          dut.io.out.bits.c.expect(expected & truncationMask)
+
+          assert(dut.io.out.bits.cmp.peek().litValue == expectedCmp.litValue)
+          dut.io.out.bits.cmp.expect(expectedCmp)
+
+          val expectedVSum = Vec.Lit(v1.zip(v2).map { case (x, y) => (x + y).U((w + 1).W) }: _*)
+
+          dut.io.out.bits.vSum.expect(expectedVSum)
+
+          val expVOutProduct = Vec.Lit(
+            v1.map { x =>
+              Vec.Lit(v2.map { y => (x * y).U((2 * w).W) }: _*)
+            }: _*
+          )
+
+          dut.io.out.bits.vOutProduct.expect(expVOutProduct)
+
+          val expectedBits = chiselTypeOf(dut.io.out.bits).Lit(
+            _.c -> (expected & truncationMask).U,
+            _.cmp -> expectedCmp,
+            _.vSum -> expectedVSum,
+            _.vDot -> v1.zip(v2).map { case (x, y) => x * y }.reduce(_ + _).U((2 * w + vecDim - 1).W),
+            _.vOutProduct -> expVOutProduct
+          )
+
+          dut.io.out.bits.expect(expectedBits)
+
+          val peekedBits = dut.io.out.bits.peek()
+          assert(peekedBits.c.litValue == expectedBits.c.litValue)
+          assert(peekedBits.cmp.litValue == expectedBits.cmp.litValue)
+
+          assert(peekedBits.elements.forall { case (name, el) => expectedBits.elements(name).litValue == el.litValue })
+
+          dut.io.out.expect(
+            chiselTypeOf(dut.io.out).Lit(
+              _.valid -> true.B,
+              _.bits -> expectedBits
+            )
+          )
+        }
+      }
+    }
+
+    it("reports failed expects correctly") {
+      val thrown = the[PeekPokeAPI.FailedExpectationException[_]] thrownBy {
+        simulate(new TestPeekPokeEnum(w)) { dut =>
+          assert(w == dut.io.in.bits.a.getWidth)
+          val vecDim = dut.vecDim
+          val truncationMask = (BigInt(1) << w) - 1
+
+          dut.io.in.bits.poke(
+            chiselTypeOf(dut.io.in.bits).Lit(
+              _.a -> 1.U,
+              _.b -> 2.U,
+              _.v1 -> Vec.Lit(Seq.fill(vecDim)(3.U(w.W)): _*),
+              _.v2 -> Vec.Lit(Seq.fill(vecDim)(4.U(w.W)): _*)
+            )
+          )
+          dut.io.in.valid.poke(true)
+          dut.io.op.poke(TestOp.Add)
+          dut.clock.step()
+
+          dut.io.out.bits.c.expect(5.U)
```

Comment:
If this is just testing that the expectation is correct, could the rest of the test be simpler?

Reply:
It is also indirectly testing `poke`s.